src/CoreShop/Component/Address/Context/RequestBased/GeoLiteBasedRequestResolver.php

```diff
@@ -25,43 +26,89 @@ final class GeoLiteBasedRequestResolver implements RequestResolverInterface
      */
     private $countryRepository;
 
+    /**
+     * @var CoreHandlerInterface
+     */
+    private $cache;
+
+    /**
+     * @var string
+     */
+    private $geoDbFile;
+
     /**
      * @param CountryRepositoryInterface $countryRepository
+     * @param CoreHandlerInterface       $cache
      */
-    public function __construct(CountryRepositoryInterface $countryRepository)
+    public function __construct(CountryRepositoryInterface $countryRepository, CoreHandlerInterface $cache)
     {
         $this->countryRepository = $countryRepository;
+        $this->cache = $cache;
+        $this->geoDbFile = PIMCORE_CONFIGURATION_DIRECTORY . '/GeoLite2-City.mmdb';
     }
 
     /**
      * {@inheritdoc}
      */
     public function findCountry(Request $request)
     {
-        $geoDbFile = PIMCORE_CONFIGURATION_DIRECTORY . '/GeoLite2-City.mmdb';
-        $record = null;
+        if (!file_exists($this->geoDbFile)) {
+            throw new CountryNotFoundException();
+        }
 
-        if (file_exists($geoDbFile)) {
-            try {
-                $reader = new Reader($geoDbFile);
+        $record = null;
+        $isoCode = null;
+        $clientIp = $request->getClientIp();
 
-                $clientIp = $request->getClientIp();
+        if ($this->checkIfIpIsPrivate($clientIp)) {
+            throw new CountryNotFoundException();
+        }
 
-                if (!$this->checkIfIpIsPrivate($clientIp)) {
-                    $record = $reader->city($clientIp);
+        $cacheKey = md5($clientIp);
 
-                    $country = $this->countryRepository->findByCode($record->country->isoCode);
+        if ($countryIsoCode = $this->cache->getItem($cacheKey)) {
+            $country = $this->countryRepository->findByCode($countryIsoCode);
 
-                    if ($country instanceof CountryInterface) {
-                        return $country;
-                    }
-                }
-            } catch (\Exception $e) {
-                //If something goes wrong, ignore the exception and throw a CountryNotFoundException
+            if ($country) {
+                return $country;
             }
         }
 
-        throw new CountryNotFoundException();
+        $countryIsoCode = $this->guessCountryByGeoLite($clientIp);
+
+        if ($countryIsoCode === null) {
+            throw new CountryNotFoundException();
+        }
+
+        $country = $this->countryRepository->findByCode($countryIsoCode);
```

Comment:
Reduce `findByCode` calls to one.

Reply:
How do you mean? introducing another function?